### ISSUE TITLE
False positive 'not-an-iterable' when using deferred initialization in a property

### ISSUE THREAD

**Issue by andy-maier** [PyCQA/pylint#4968]
### Bug description

```python
#!/usr/bin/env python
# File: pylint_not_an_iterable.py
"""
Demonstrate the false positive error 'not-an-iterable' in Pylint.

Pylint 2.10.2 raises the error on Python 3.8 and 3.9, but not on Python
3.6 and 3.7.

In the original code, the error was raised only with astroid 2.7.3 and not
with astroid 2.7.2. This simplified code here raises the error with both
astroid versions.

The deferred initialization in NamedObject.names() is key for causing the
Pylint error to be raised; when performing the initialization in
NamedObject.__init__(), Pylint does not raise the error.
"""

# pylint: disable=missing-class-docstring,missing-function-docstring
# pylint: disable=useless-object-inheritance,too-few-public-methods
# pylint: disable=raise-missing-from,invalid-name

from __future__ import print_function


class NamedObject(object):

    def __init__(self, names):
        self._org_names = names
        self._names = None

    def _determine_names(self):
        self._names = [n.lower() for n in self._org_names]

    @property
    def names(self):
        # Deferred initialization
        if self._names is None:
            self._determine_names()
        return self._names


obj1 = NamedObject(['Bla'])
names1 = getattr(obj1, 'names')

# Pylint raises error 'not-an-iterable' for names1 on the following line:
names1_u = [n.upper() for n in names1]

print(names1_u)
```

### Command used

```shell
pylint pylint_not_an_iterable.py  # with no config file present
```

### Pylint output

```shell
On Python 3.9 with astroid 2.7.2 and with 2.7.3:
************* Module pylint_not_an_iterable
pylint_not_an_iterable.py:46:31: E1133: Non-iterable value names1 is used in an iterating context (not-an-iterable)
```

### Expected behavior

Since the value returned from the `names` property is an iterable, Pylint should not raise the "not-an-iterable" error.

Pylint probably does not figure out that the initialization in `NamedObject.names()` is deferred. When initializing in the init method instead, Pylint does not raise the error.

Also, there seems to be a potential dependency to the astroid version: The original code did not cause this error to be raised when using astroid 2.7.2 on Python 3.8 and 3.9 (see [these test runs from 8/29](https://github.com/pywbem/pywbem/actions/runs/1178348129)), but did cause this error to be raised with astroid 2.7.3 on Python 3.8 and 3.9 (see [these test runs from 9/5](https://github.com/pywbem/pywbem/actions/runs/1202131339)). astroid 2.7.3 was released on 8/30, i.e. in between these two sets of runs. However, the simplified code shown above causes the error to be raised for both astroid versions on Python 3.9. I did not dig deeper into the reasons for what triggers the version dependency.

### Pylint version

```shell
On Python 3.9 with astroid 2.7.3:
pylint 2.10.2
astroid 2.7.3  # and also with 2.7.2, see above.
Python 3.9.6 (default, Jun 29 2021, 06:20:32) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

macOS 10.15.7

### Additional dependencies

Dependencies of Pylint:

On Python 3.9 with astroid 2.7.3:
$ pipdeptree --python $(which python) -p pylint
pylint==2.10.2
  - astroid [required: >=2.7.2,<2.8, installed: 2.7.3]
    - lazy-object-proxy [required: >=1.4.0, installed: 1.6.0]
    - setuptools [required: >=20.0, installed: 57.5.0]
    - wrapt [required: >=1.11,<1.13, installed: 1.12.1]
  - isort [required: >=4.2.5,<6, installed: 5.9.3]
  - mccabe [required: >=0.6,<0.7, installed: 0.6.1]
  - platformdirs [required: >=2.2.0, installed: 2.3.0]
  - toml [required: >=0.7.1, installed: 0.10.2]

On Python 3.9 with astroid 2.7.2:
$ pipdeptree --python $(which python) -p pylint
pylint==2.10.2
  - astroid [required: >=2.7.2,<2.8, installed: 2.7.2]
    - lazy-object-proxy [required: >=1.4.0, installed: 1.6.0]
    - setuptools [required: >=20.0, installed: 57.5.0]
    - wrapt [required: >=1.11,<1.13, installed: 1.12.1]
  - isort [required: >=4.2.5,<6, installed: 5.9.3]
  - mccabe [required: >=0.6,<0.7, installed: 0.6.1]
  - platformdirs [required: >=2.2.0, installed: 2.3.0]
  - toml [required: >=0.7.1, installed: 0.10.2]

**Comment by vtgn** [comment]
Same problem as in https://github.com/pylint-dev/pylint/issues/701.
Like I just said in my comment in this 701 issue, Pylint is not a static typing linter and should better delete the `not-an-iterable` rule.
The static typing linters like Mypy and Pylance notify no error in the code of @andy-maier above, because there are no info about the typing of the objects (they use `Any` for typing `names1`). Wanting to continue to check and fix this `not-an-iterable` rule is a very hard task that even the static typing linters don't do.
